### ISSUE TITLE
typescript/consistent-type-imports doesn't respect tsconfig

### ISSUE THREAD

**Issue by anotheri** [oxc-project/oxc#13609]
there is a problem with `"typescript/consistent-type-imports": "error"` rule in `.oxlintrc.json` (i tried different options setup and it doesn't really matter) in nestjs project, oxlint throws the error on import decorated classes and tries to convert them to types, although it's incorrect.

I've found this related issue https://github.com/typescript-eslint/typescript-eslint/issues/10200, it works with eslint as expected.
i've also tried different options in `tsconfig.json` but my currently working one looks like this:

```
{
  "compilerOptions": {
    "module": "esnext",
    "declaration": true,
    "removeComments": true,
    "isolatedModules": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "resolveJsonModule": true,
    "target": "es2023",
    "sourceMap": true,
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true,
    "skipLibCheck": true,
    "strict": true,
    "moduleResolution": "bundler",
    "paths": {
      "@src/*": ["src/*"],
      "@tests/*": ["tests/*"]
    },
    "forceConsistentCasingInFileNames": true,
    "noUncheckedIndexedAccess": true
  },
  "include": ["**/*.ts", "vitest.config.mts"]
}
```

**Comment by camc314** [comment]
can you share a code repo with the diagnostic and full code? 

without this, it's difficult to identify what's going on.

thank you

**Comment by anotheri** [comment]
@camc314 here you are: https://github.com/anotheri/nestjs-oxlint
if you run `yarn test` all tests pass, but if you run `yarn lint` it'll throw oxlint errors related related to issue i've described above.
on other hand, if you fix those oxlint errors with `yarn lint:fix`, some of the tests start to fail.

**Comment by camc314** [comment]
@Dunqing do you have any ideas here as to the best solution? I guess depending on the tsconfig, semantic needs to update the reference flags differently?

**Comment by Dunqing** [comment]
> do you have any ideas here as to the best solution? I guess depending on the tsconfig, semantic needs to update the reference flags differently?

The current references flags are correct. I think this is a footgun if we change reference flags according to the `emitDecorataMetadata` option, because it may break other rules, and it isn't worth doing it for a special single rule.

I thought of two possible solutions to address this:

1. Don't report errors if the import is used as a type annotation of `param`, `method`, and `property` that includes decorators. (Needs to traverse upward to check)
2. Same as `typescript-eslint`, disable this rule if `experimentalDecorators` and `emitDecoratorMetadata` are enabled.

The first option is preferred because it still keeps working on the non-decorator type annotations, but it may affect performance a little bit.

**Comment by anotheri** [comment]
just a reminder, it's still an issue with 
- `oxlint-tsgolint: ^0.9.2`
- `oxlint: ^1.34.0`

**Comment by serega3000** [comment]
i used this temporary workaround to fix my codebase:
https://gist.github.com/serega3000/08e276066ea9042be139211aa3263492

this script:
1. search for classes with decorators in directory
2. remove type imports for class constructors
3. add ignore comment to remove oxlint error

**Comment by anotheri** [comment]
Thanks for shared workaround! Although I'd better stick with eslint till it's properly addressed in oxlint rather than bypass these rules

**Comment by nikolailehbrink** [comment]
Was just coming across this and wanted to bump it. Also an issue for us.

**Comment by thiagomini** [comment]
I think oxlint should apply the same logic as the [ESLint counterpart](https://typescript-eslint.io/rules/consistent-type-imports/#caveat-decorators--experimentaldecorators-true--emitdecoratormetadata-true) and just don't report errors when the experimental decorators is true and there are decorators in the file. I'm postponing migrating our NestJS app to oxlint specifically because of this rule.

**Comment by ianpardal** [comment]
Any update here? As @thiagomini mentioned, ensuring the logic doesn't run on files with experimental decorators in them would be the easiest/fastest unblocking solution. Any project that extensively uses experimental decorators (i.e. every NestJS project) currently isn't compatible with this rule.

I can remove this rule altogether from my project, but it still would be nice to have the rule for files that don't contain them!

**Comment by therockstorm** [comment]
`typescript/consistent-type-imports` still reports imports that are only syntactically used in decorated constructor parameter types, even when the resolved tsconfig has both:

```json
{
  "experimentalDecorators": true,
  "emitDecoratorMetadata": true
}
```

Applying the autofix changes the emitted decorator metadata for class constructor injection:

- `tsc` 5.8.3 and 6.0.3: `design:paramtypes` changes from `[ChargeService]` to `[Function]`.
- SWC 1.15.41 with `legacyDecorator: true` and `decoratorMetadata: true`: metadata changes from a runtime import reference to an unbound local fallback that evaluates to `Object`.

Either change breaks NestJS-style class-token dependency injection because the reflected constructor token is no longer the imported class.

This is the runtime breakage described in the typescript-eslint post "Changes to `consistent-type-imports` with Legacy Decorators and Decorator Metadata": when `experimentalDecorators` and `emitDecoratorMetadata` are both enabled, typescript-eslint does not report errors in files containing decorators.

## Repro

This repro is dependency-free; the decorator is a local stub. The important part is that the decorator flags are present through a tsconfig `extends` chain.

`tsconfig.base.json`

```json
{
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "strict": true,
    "target": "ES2022",
    "module": "CommonJS"
  }
}
```

`tsconfig.json`

```json
{
  "extends": "./tsconfig.base.json",
  "include": ["src/**/*.ts"]
}
```

`src/decorators.ts`

```ts
export function Injectable(): ClassDecorator {
  return function decorateClass(): void {};
}
```

`src/charge.service.ts`

```ts
export class ChargeService {
  public toString(): string {
    return "charge-service";
  }
}
```

`src/example.service.ts`

```ts
import { Injectable } from "./decorators";
import { ChargeService } from "./charge.service";

@Injectable()
export class ExampleService {
  public constructor(private readonly chargeService: ChargeService) {}

  public run(): void {
    this.chargeService.toString();
  }
}
```

Commands:

```sh
npx oxlint@1.69.0 --tsconfig tsconfig.json -D typescript/consistent-type-imports src/example.service.ts
npx -p oxlint@1.69.0 -p oxlint-tsgolint@0.23.0 oxlint --type-aware --tsconfig tsconfig.json -D typescript/consistent-type-imports src/example.service.ts
```

Actual in both modes:

```text
src/example.service.ts:2:1: error typescript(consistent-type-imports): All imports in the declaration are only used as types. Use `import type`.
```

The autofix changes the import to:

```ts
import type { ChargeService } from "./charge.service";
```

Expected:

When the resolved tsconfig has `experimentalDecorators: true` and `emitDecoratorMetadata: true`, this rule should not report this import in a file containing decorators, or should otherwise avoid fixes that alter decorator metadata. That would match typescript-eslint's documented compromise.

## Verification matrix

- `oxlint@1.60.0`, plain `--tsconfig`: reproduces.
- `oxlint@1.60.0` with `--type-aware` and `oxlint-tsgolint@0.23.0`: reproduces.
- `oxlint@1.68.0`, plain `--tsconfig`: reproduces.
- `oxlint@1.68.0` with `--type-aware` and `oxlint-tsgolint@0.23.0`: reproduces.
- `oxlint@1.69.0` (npm latest on 2026-06-12; GitHub release `apps_v1.69.0` was published 2026-06-08), plain `--tsconfig`: reproduces.
- `oxlint@1.69.0` with `--type-aware` and `oxlint-tsgolint@0.23.0`: reproduces.
- Current `main` at `f57b55db11`: source-inspected. I could not execute the local binary because this machine has no installed Rust toolchain and rustup cannot download Rust 1.96.0 from `static.rust-lang.org` in the current network environment. Static inspection shows the rule body still has no decorator/tsconfig guard and the decorator-metadata test cases are still commented out.

## Runtime impact

With the value import, TypeScript emits the constructor token:

```js
__metadata("design:paramtypes", [charge_service_1.ChargeService])
```

After the autofix to `import type`, TypeScript emits:

```js
__metadata("design:paramtypes", [Function])
```

With SWC 1.15.41 and:

```json
{
  "jsc": {
    "parser": { "syntax": "typescript", "decorators": true },
    "target": "es2022",
    "transform": { "legacyDecorator": true, "decoratorMetadata": true }
  },
  "module": { "type": "commonjs" }
}
```

the value import emits:

```js
_ts_metadata("design:paramtypes", [
  typeof _chargeservice.ChargeService === "undefined" ? Object : _chargeservice.ChargeService
])
```

After the autofix, SWC emits:

```js
_ts_metadata("design:paramtypes", [
  typeof ChargeService === "undefined" ? Object : ChargeService
])
```

Because `ChargeService` is no longer imported, this evaluates to `Object`, not the class constructor token.

## Root cause in oxc

The current rule classifies an import as type-only if every semantic reference for that import satisfies `Reference::is_type()`:

- Release `apps_v1.69.0`: `crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs`, lines 224-318 and helper at lines 355-364.
- Current `main` (`f57b55db11`): same structure.

That is insufficient for legacy decorator metadata because the runtime reference is implicit in compiler emit; it is not a normal value reference in the source AST.

The rule config only exposes `disallowTypeAnnotations`, `fixStyle`, and `prefer`; none of those can express `experimentalDecorators` or `emitDecoratorMetadata`.

There is also an explicit TODO in the rule tests:

```rust
// TODO: To support decorator in this rule, need <https://github.com/oxc-project/oxc/pull/3645>
// experimentalDecorators: true + emitDecoratorMetadata: true
```

The commented-out fixtures include exactly the missing shape:

```ts
import Foo from 'foo';
@deco
class A {
  constructor(foo: Foo) {}
}
```

The structural gap appears to be that `--tsconfig` is only carried into linter service resolution, not into rule context as compiler options:

- `apps/oxlint/src/lint.rs` passes the path into `LintServiceOptions::with_tsconfig`.
- `crates/oxc_linter/src/service/mod.rs` describes that field as being for path aliases and project references.
- `crates/oxc_linter/src/service/runtime.rs` consumes it in `get_resolver`.
- `LintContext` / `ContextHost` do not expose resolved TypeScript compiler options to this rule.

One confusing extra detail: the current Oxlint docs for `prefer: "type-imports"` say the rule enforces type imports "except referenced by metadata of decorators", but release 1.69.0 and current source do not appear to implement that exception for this case.

## Fix sketch

The safest parity fix is the typescript-eslint compromise:

1. Resolve the relevant tsconfig for the file, including `extends`.
2. Expose at least `experimentalDecorators` and `emitDecoratorMetadata` to lint rules.
3. In `typescript/consistent-type-imports`, if both options are true and the file contains decorators, suppress reporting/fixing for that file.

This avoids unsafe autofixes in NestJS, TypeORM, TypeGraphQL, class-validator, and similar legacy-decorator metadata code.

A more precise but riskier fix is the option discussed above: do not change global semantic reference flags, but have this rule treat imports used in decorated constructor parameters, method parameters/returns, and decorated properties as not safely type-only. That would preserve reports in more non-decorator locations, but needs careful ancestor checks and test coverage for class, constructor parameter, method, accessor, and property decorators.

## References

- Related DI issue: https://github.com/oxc-project/oxc/issues/20433
- Closed parser-options attempt: https://github.com/oxc-project/oxc/pull/3645
- Latest tested release: https://github.com/oxc-project/oxc/releases/tag/apps_v1.69.0
- Release source, rule body: https://github.com/oxc-project/oxc/blob/apps_v1.69.0/crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs#L192-L318
- Release source, TODO fixtures: https://github.com/oxc-project/oxc/blob/apps_v1.69.0/crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs#L1385-L1396
- typescript-eslint behavior: https://typescript-eslint.io/blog/changes-to-consistent-type-imports-with-decorators/
- Oxlint docs currently claiming the decorator-metadata exception: https://oxc.rs/docs/guide/usage/linter/rules/typescript/consistent-type-imports

## Disclosure

This was prepared with AI assistance and has been reviewed by a human.

**Comment by matthias-feddersen** [comment]
To fix this in my nestjs project, since the rule won't change I used @Inject( to inject the dependencies which is the intended way to do it explicitly, then it sees it as a class and doesnt want to change it to import type

**Comment by stavalfi-oasis** [comment]
any progress? and pls dont make what eslint-typescript did: https://typescript-eslint.io/blog/changes-to-consistent-type-imports-with-decorators/

we need an accurate solution when:
```json
 "emitDecoratorMetadata": true,
 "experimentalDecorators": true
```

**Comment by prandogabriel** [comment]
Hit this exact issue independently on a large NestJS codebase (oxlint 1.75.0). Concrete repro shape:

```ts
// external.service.ts
export class ExternalService {}

// base.repository.ts
export abstract class BaseRepository<T> {
  constructor(protected readonly externalService: ExternalService) {}
}

// item.repository.ts
@Injectable()
export class ItemRepository extends BaseRepository<Item> {
  constructor(externalService: ExternalService) {
    super(externalService);
  }
}
```

`oxlint --fix` converts the `ExternalService` import in `item.repository.ts` to `import type` (nothing else in the file uses it as a value). Combined with `emitDecoratorMetadata: true`, this makes Nest's `Test.createTestingModule` resolve `externalService` as `undefined` at runtime instead of the injected instance — no compile error, no thrown error, just silent `undefined`. Confirmed with a real Jest DI test before/after the autofix, and separately by running it across a full test suite: hundreds of tests failed once the fix was applied broadly.

Worth broadening the scope of this issue: **`typescript/no-useless-constructor`'s autofix has the identical blind spot**, on the same shape of code:

```ts
@Injectable()
export class ItemRepository extends BaseRepository<Item> {
  constructor(externalService: ExternalService) {
    super(externalService);
  }
}
```

`no-useless-constructor` correctly flags this constructor as removable under plain JS/TS semantics (it just forwards to `super` with the same args). But deleting it breaks DI the same way — verified empirically with a Jest test: after removing the constructor, `externalService` resolves to `undefined`, not even a throw, because a subclass with no explicit constructor gets no `design:paramtypes` metadata emitted for it under `emitDecoratorMetadata`, even though the base class's own constructor is unchanged.

Both rules' fixers appear to reason purely on static type/usage analysis without modeling that a class carrying a decorator that triggers `emitDecoratorMetadata` needs its own constructor's parameter types preserved as real values — regardless of whether the linter calls the import "type-only" or the constructor "useless" in vanilla JS terms. A shared guard (skip autofix/suggestion for any class carrying a decorator, or any class extending a base with its own constructor) might be more robust than fixing this per-rule, since it's the same root cause surfacing in at least two different rules now.

**Comment by sergio-milu** [comment]
NestJS 12 was released with [oxlint as default linter](https://docs.nestjs.com/migration-guide#linting-defaults), but this rule is a hard-stopper for decorators, any updates in this regard?

**Comment by JSteunou** [comment]
@Dunqing you seem to have some idea to solve this but this issue was assigned to @camc314 without activities so far...

I also postponed our migration from ESLint because of decorator usage in NestJS.

A year later, here we are, how can we help?

I consider myself a good developer but with no skills in Rust, do you accept AI generated PR? It is the best I could offer to unlock this situation.

**Comment by Matsuuu** [comment]
Seconding @JSteunou 's recommendation. A lot of projects are currently migrating towards OxLint and with this rule being a big part of large typescript decorator-driven projects, having this rule fixed would be heavily appriciated
